src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Formats.Cbor;
+
+namespace System.Security.Cryptography.Cose
+{
+    public class CoseHeaderMap : IEnumerable<(CoseHeaderLabel, ReadOnlyMemory<byte>)>
+    {
+        private static readonly byte[] s_emptyBstrEncoded = new byte[] { 0x40 };
+        private bool _isReadOnly;
+        public bool IsReadOnly { get => _isReadOnly; internal set => _isReadOnly = value; }
+        private readonly Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>> _headerParameters = new Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>>();
+
+        public CoseHeaderMap() : this (isReadOnly: false) { }
+
+        internal CoseHeaderMap(bool isReadOnly)
+        {
+            _isReadOnly = isReadOnly;
+        }
+
+        public bool TryGetEncodedValue(CoseHeaderLabel label, out ReadOnlyMemory<byte> encodedValue)
+            => _headerParameters.TryGetValue(label, out encodedValue);
+
+        public ReadOnlyMemory<byte> GetEncodedValue(CoseHeaderLabel label)
+            => _headerParameters[label];
+
+        public int GetValueAsInt32(CoseHeaderLabel label)
+        {
+            var reader = new CborReader(_headerParameters[label]);
+            int retVal = reader.ReadInt32();
+            Debug.Assert(reader.BytesRemaining == 0);
+            return retVal;
+        }
+
+        public string GetValueAsString(CoseHeaderLabel label)
+        {
+            var reader = new CborReader(_headerParameters[label]);
+            string retVal = reader.ReadTextString();
+            Debug.Assert(reader.BytesRemaining == 0);
+            return retVal;
+        }
+
+        public ReadOnlySpan<byte> GetValueAsBytes(CoseHeaderLabel label)
+        {
+            var reader = new CborReader(_headerParameters[label]);
+            ReadOnlySpan<byte> retVal = reader.ReadByteString();
+            Debug.Assert(reader.BytesRemaining == 0);
+            return retVal;
+        }
+
+        public void SetEncodedValue(CoseHeaderLabel label, ReadOnlyMemory<byte> encodedValue)
```

Comment:
Do we want ReadOnlyMemory here, or ReadOnlySpan?


With ReadOnlyMemory the caller has the ability to change the value after ValidateHeaderValue runs, but before encoding happens.

With ReadOnlySpan we have to make a copy, but that removes the potential ToC/ToU error.

For a string input, which is a nearly-immutable type, we're already making a copy (by running it through the CBOR encoder), so maybe the public member should be ReadOnlySpan.